### ISSUE TITLE
setup confuses packages depending on order in install_requires list

### ISSUE THREAD

**Issue by bb-migration** [pypa/setuptools#498]
Originally reported by: **ikseek (Bitbucket: [ikseek](http://bitbucket.org/ikseek), GitHub: [ikseek](http://github.com/ikseek))**

---

Try this simple reproducer with python3.5.

```
#!python

from setuptools import setup, find_packages

INSTALL_REQUIRES = [
    'Flask',
    'flask-mongoengine',
]

setup(
    include_package_data=True,
    author='Author',
    author_email='author@email.com',
    url='https://url.com',
    install_requires=INSTALL_REQUIRES,
    license='PROPRIETARY',
    long_description='Long description',
    name='pip-requirements-order-bug-reproducer',
    packages=find_packages(),
    version='1'
)

```

**Observed behavior:**
Running  "python3 setup.py develop" fails to install Flask, because it tries to install flask-mongoengine second time!

**Expected behavior:**
Running  "python3 setup.py develop" should install both Flask and flask-mongoengine

**Note:**
Swapping entries in INSTALL_REQUIRES list hides the issue.

---
- Bitbucket: https://bitbucket.org/pypa/setuptools/issue/498

**Comment by ngeraedts** [comment]
Can confirm this is an issue on python 2.7.14 as well.

```
from setuptools import setup

setup(
    name="install_requires_order_test",
    install_requires = [
        "flake8",
        "flake8-builtins",
    ]
)
```

As the original poster mentioned, reversing the definition order results in the dependencies being correctly installed.

**Comment by jaraco** [comment]
This issue is a duplicate of #196.

**Comment by jaraco** [comment]
I should point out that it's somewhat unexpected that you would be encountering this issue with `install_requires` directives, even with Setuptools prior to 36.4.0, unless you're using `easy_install` to install packages, but the PyPA recommends using pip to install packages, which isn't subject to that issue.

**Comment by benoit-pierre** [comment]
I don't think it's a duplicate, as I can reproduce the issue. The fix for #196 does not work for a requirement dependencies.

**Comment by jaraco** [comment]
Oh, interesting... and surprising.

**Comment by cas--** [comment]
Yep definitely an `install_requires` ordering issue causing service-identity to fail installing in #1510:

This example fails with `The 'pyasn1' distribution was not found and is required`:
```python
from setuptools import setup
setup(name='test', install_requires=['pyasn1', 'pyasn1-modules'])
```

Changing the order works however:
```python
from setuptools import setup
setup(name='test', install_requires=['pyasn1-modules', 'pyasn1'])
```

**Comment by looking-for-a-job** [comment]
I made a small package that will help in the simplest cases  https://pypi.org/project/requirements-order/

**Comment by Zethson** [comment]
This issue is still persistent.

**Comment by jastip** [comment]
> This issue is still persistent.

On my side, this issue showed up because I install pip using easy_install, after I uninstall pip using 

$ easy_install uninstall pip

the problem gone

**Comment by ryan-collingham** [comment]
I'm also seeing this issue in setuptools 41.0.0 with python 3.7.3, when installing a package that depends on both "Flask" and "flask_restplus".

I only see the issue if I install with `python setup.py install`, if I do `pip install .` instead there is no issue.

**Comment by avbentem** [comment]
For future readers/googlers, as documented in a few other places:

Specifying an explicit version may work too. (This does not need to be some previous version, though some may get confused thinking that using some older version of a dependency solved their problem.)

When using `setup.cfg`, fails (for me):

```
install_requires =
    ...
    Flask
    Flask-SQLAlchemy
    ...
    apispec[validation]
    apispec-webframeworks
```

> error: The 'Flask' distribution was not found and is required by [...]
> error: The 'apispec[validation]' distribution was not found and is required by [...]

Works (for me):

```
install_requires =
    ...
    Flask>=1.1.2
    Flask-SQLAlchemy
    ...
    apispec[validation]>=v4.4.0b1
    apispec-webframeworks
```

Changing the order indeed works as well (for me):

```
install_requires =
    ...
    Flask-SQLAlchemy
    Flask
    ...
    apispec-webframeworks
    apispec[validation]
```

Aside, it seems that running `python setup.py develop` multiple times resolves this too.
